doc/guides/design_guidelines.rst

```diff
@@ -0,0 +1,77 @@
+.. _design_guidelines:
+
+API Design Guidelines
+#####################
+
+Zephyr development and evolution is a group effort, and to simplify
+maintenance and enhancements there are some general policies that should
+be followed when developing a new capability or interface.
+
+Using Callbacks
+***************
+
+Many APIs involve passing a callback as a parameter or as a member of a
+configuration structure.  The following policies should be followed when
+specifying the signature of a callback:
+
+* The first parameter should be a pointer to the object most closely
+  associated with the callback.  In the case of device drivers this
+  would be ``struct device *dev``.  For library functions it may be a
+  pointer to another object that was referenced when the callback was
+  provided.
+
+* The next parameter(s) should be additional information specific to the
+  callback invocation, such as a channel identifier, new status value,
+  and/or a message pointer followed by the message length.
+
+* The final parameter should be a ``void *user_data`` pointer carrying
+  context that allows a shared callback function to locate additional
+  material necessary to process the callback.
+
+An exception to providing ``user_data`` as the last parameter may be
```

Comment:
would be great if you added a few examples to illustrate that with some existing or a dummy API...

Reply:
Done, below